[BUG]: non-pythonic function signature for native_enum

### Required prerequisites

- [x] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [x] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [x] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### What version (or hash if on master) of pybind11 are you using?

master

### Problem description

I tested using pybind11-stubgen with the latest pybind11 master. When the `native_enum` is used in function signatures, pybind11-stubgen cannot "link" the argument to the enum from the module (see this [issue](https://github.com/sizmailov/pybind11-stubgen/issues/249)). To test if something is "wrong" in pybind11 I wrote following test:

```cpp
#include <pybind11/native_enum.h>

namespace test {
    enum class fake{ x };
    enum class native{ x };
}

TEST_SUBMODULE(docstring_options, m) {
    // test_docstring_options
    {
        py::enum_<test::fake>(m, "Fake").value("x", test::fake::x);
        py::native_enum<test::native>(m, "Native", "enum.Enum").value("x", test::native::x).finalize();

        options.enable_function_signatures();

        m.def("test_function_enum", [](test::fake, test::native){}, py::arg("f"), py::arg("n"));
    }
}
```

and

```py
    assert m.test_function_enum.__doc__.starts-with(
        "test_function_enum(f: pybind11_tests.docstring_options.Fake, n: pybind11_tests.docstring_options.Native) -> None"
    )
```

This fails because the signature of the second argument is "C++-like", i.e. `test::native` instead of `pybind11_tests.docstring_options.Native`.

```sh
E       AssertionError: assert False
E        +  where False = <built-in method startswith of str object at 0x1179b92c0>('test_function_enum(f: pybind11_tests.docstring_options.Fake, n: pybind11_tests.docstring_options.Native) -> None')
E        +    where <built-in method startswith of str object at 0x1179b92c0> = 'test_function_enum(f: pybind11_tests.docstring_options.Fake, n: test::native) -> None\n'.startswith
E        +      where 'test_function_enum(f: pybind11_tests.docstring_options.Fake, n: test::native) -> None\n' = <built-in method test_function_enum of pybind11_detail_function_record_v1_system_libcpp_abi1 object at 0x105b20990>.__doc__
E        +        where <built-in method test_function_enum of pybind11_detail_function_record_v1_system_libcpp_abi1 object at 0x105b20990> = m.test_function_enum


../../tests/test_docstring_options.py:31: AssertionError
```

### Reproducible example code

```text

```

### Is this a regression? Put the last known working version here if it is.

Not a regression

--- a/include/pybind11/detail/native_enum_data.h
+++ b/include/pybind11/detail/native_enum_data.h
@@ -182,6 +182,10 @@ inline void native_enum_data::finalize() {
     if (module_name) {
         py_enum.attr("__module__") = module_name;
     }
+    if (hasattr(parent_scope, "__qualname__")) {
+        const auto parent_qualname = parent_scope.attr("__qualname__").cast<std::string>();
+        py_enum.attr("__qualname__") = str(parent_qualname + "." + enum_name.cast<std::string>());
+    }
     parent_scope.attr(enum_name) = py_enum;
     if (export_values_flag) {
         for (auto member : members) {

--- a/include/pybind11/pybind11.h
+++ b/include/pybind11/pybind11.h
@@ -160,6 +160,9 @@ inline std::string generate_function_signature(const char *type_caster_name_fiel
                 handle th((PyObject *) tinfo->type);
                 signature += th.attr("__module__").cast<std::string>() + "."
                              + th.attr("__qualname__").cast<std::string>();
+            } else if (auto th = detail::global_internals_native_enum_type_map_get_item(*t)) {
+                signature += th.attr("__module__").cast<std::string>() + "."
+                             + th.attr("__qualname__").cast<std::string>();
             } else if (func_rec->is_new_style_constructor && arg_index == 0) {
                 // A new-style `__init__` takes `self` as `value_and_holder`.
                 // Rewrite it to the proper class type.

--- a/tests/test_native_enum.cpp
+++ b/tests/test_native_enum.cpp
@@ -29,6 +29,8 @@ enum class member_doc { mem0, mem1, mem2 };
 
 struct class_with_enum {
     enum class in_class { one, two };
+
+    in_class value = in_class::one;
 };
 
 // https://github.com/protocolbuffers/protobuf/blob/d70b5c5156858132decfdbae0a1103e6a5cb1345/src/google/protobuf/generated_enum_util.h#L52-L53
@@ -125,6 +127,8 @@ TEST_SUBMODULE(native_enum, m) {
         .value("one", class_with_enum::in_class::one)
         .value("two", class_with_enum::in_class::two)
         .finalize();
+
+    py_class_with_enum.def(py::init()).def_readwrite("value", &class_with_enum::value);
 
     py::native_enum<native>(m, "Native", "enum.Enum").value("x", native::x).finalize();
 

--- a/tests/test_native_enum.cpp
+++ b/tests/test_native_enum.cpp
@@ -30,7 +30,7 @@ enum class member_doc { mem0, mem1, mem2 };
 struct class_with_enum {
     enum class in_class { one, two };
 
-    in_class value = in_class::one;
+    in_class nested_value = in_class::one;
 };
 
 // https://github.com/protocolbuffers/protobuf/blob/d70b5c5156858132decfdbae0a1103e6a5cb1345/src/google/protobuf/generated_enum_util.h#L52-L53
@@ -42,7 +42,7 @@ enum some_proto_enum : int { Zero, One };
 template <>
 struct is_proto_enum<some_proto_enum> : std::true_type {};
 
-enum class native { x };
+enum class func_sig_rendering {};
 
 } // namespace test_native_enum
 
@@ -128,14 +128,19 @@ TEST_SUBMODULE(native_enum, m) {
         .value("two", class_with_enum::in_class::two)
         .finalize();
 
-    py_class_with_enum.def(py::init()).def_readwrite("value", &class_with_enum::value);
-
-    py::native_enum<native>(m, "Native", "enum.Enum").value("x", native::x).finalize();
+    py_class_with_enum.def(py::init())
+        .def_readwrite("nested_value", &class_with_enum::nested_value);
 
     m.def("isinstance_color", [](const py::object &obj) { return py::isinstance<color>(obj); });
 
     m.def("pass_color", [](color e) { return static_cast<int>(e); });
     m.def("return_color", [](int i) { return static_cast<color>(i); });
+
+    py::native_enum<func_sig_rendering>(m, "func_sig_rendering", "enum.Enum").finalize();
+    m.def(
+        "pass_and_return_func_sig_rendering",
+        [](func_sig_rendering e) { return e; },
+        py::arg("e"));
 
     m.def("pass_some_proto_enum", [](some_proto_enum) { return py::none(); });
     m.def("return_some_proto_enum", []() { return some_proto_enum::Zero; });
@@ -241,6 +246,4 @@ TEST_SUBMODULE(native_enum, m) {
 #else
     m.attr("native_enum_missing_finalize_failure") = "For local testing only: terminates process";
 #endif
-
-    m.def("function_with_native_enum", [](native n) { return n; }, py::arg("n"));
 }

--- a/tests/test_native_enum.cpp
+++ b/tests/test_native_enum.cpp
@@ -40,6 +40,8 @@ enum some_proto_enum : int { Zero, One };
 template <>
 struct is_proto_enum<some_proto_enum> : std::true_type {};
 
+enum class native { x };
+
 } // namespace test_native_enum
 
 PYBIND11_NAMESPACE_BEGIN(PYBIND11_NAMESPACE)
@@ -123,6 +125,8 @@ TEST_SUBMODULE(native_enum, m) {
         .value("one", class_with_enum::in_class::one)
         .value("two", class_with_enum::in_class::two)
         .finalize();
+
+    py::native_enum<native>(m, "Native", "enum.Enum").value("x", native::x).finalize();
 
     m.def("isinstance_color", [](const py::object &obj) { return py::isinstance<color>(obj); });
 
@@ -233,4 +237,6 @@ TEST_SUBMODULE(native_enum, m) {
 #else
     m.attr("native_enum_missing_finalize_failure") = "For local testing only: terminates process";
 #endif
+
+    m.def("function_with_native_enum", [](native n) { return n; }, py::arg("n"));
 }

--- a/tests/test_native_enum.py
+++ b/tests/test_native_enum.py
@@ -84,15 +84,10 @@ def test_enum_members(enum_type, members):
 def test_pickle_roundtrip(enum_type, members):
     for name, _ in members:
         orig = enum_type[name]
-        if enum_type is m.class_with_enum.in_class:
-            # This is a general pickle limitation.
-            with pytest.raises(pickle.PicklingError):
-                pickle.dumps(orig)
-        else:
-            # This only works if __module__ is correct.
-            serialized = pickle.dumps(orig)
-            restored = pickle.loads(serialized)
-            assert restored == orig
+        # This only works if __module__ is correct.
+        serialized = pickle.dumps(orig)
+        restored = pickle.loads(serialized)
+        assert restored == orig
 
 
 @pytest.mark.parametrize("enum_type", [m.flags_uchar, m.flags_uint])
@@ -301,6 +296,14 @@ def test_native_enum_missing_finalize_failure():
     if not isinstance(m.native_enum_missing_finalize_failure, str):
         m.native_enum_missing_finalize_failure()
         pytest.fail("Process termination expected.")
+
+
+def test_property_type_hint():
+    prop = m.class_with_enum.__dict__["value"]
+    assert isinstance(prop, property)
+    assert prop.fget.__doc__.startswith(
+        "(self: pybind11_tests.native_enum.class_with_enum) -> pybind11_tests.native_enum.class_with_enum.in_class"
+    )
 
 
 def test_function_signature():

--- a/tests/test_native_enum.py
+++ b/tests/test_native_enum.py
@@ -54,6 +54,8 @@ MEMBER_DOC_MEMBERS = (
     ("mem2", 2),
 )
 
+FUNC_SIG_RENDERING_MEMBERS = ()
+
 ENUM_TYPES_AND_MEMBERS = (
     (m.smallenum, SMALLENUM_MEMBERS),
     (m.color, COLOR_MEMBERS),
@@ -62,6 +64,7 @@ ENUM_TYPES_AND_MEMBERS = (
     (m.flags_uint, FLAGS_UINT_MEMBERS),
     (m.export_values, EXPORT_VALUES_MEMBERS),
     (m.member_doc, MEMBER_DOC_MEMBERS),
+    (m.func_sig_rendering, FUNC_SIG_RENDERING_MEMBERS),
     (m.class_with_enum.in_class, CLASS_WITH_ENUM_IN_CLASS_MEMBERS),
 )
 
@@ -148,6 +151,22 @@ def test_return_color_fail():
     with pytest.raises(ValueError) as excinfo_cast:
         m.return_color(2)
     assert str(excinfo_cast.value) == str(excinfo_direct.value)
+
+
+def test_property_type_hint():
+    prop = m.class_with_enum.__dict__["nested_value"]
+    assert isinstance(prop, property)
+    assert prop.fget.__doc__.startswith(
+        "(self: pybind11_tests.native_enum.class_with_enum)"
+        " -> pybind11_tests.native_enum.class_with_enum.in_class"
+    )
+
+
+def test_func_sig_rendering():
+    assert m.pass_and_return_func_sig_rendering.__doc__.startswith(
+        "pass_and_return_func_sig_rendering(e: pybind11_tests.native_enum.func_sig_rendering)"
+        " -> pybind11_tests.native_enum.func_sig_rendering"
+    )
 
 
 def test_type_caster_enum_type_enabled_false():
@@ -296,17 +315,3 @@ def test_native_enum_missing_finalize_failure():
     if not isinstance(m.native_enum_missing_finalize_failure, str):
         m.native_enum_missing_finalize_failure()
         pytest.fail("Process termination expected.")
-
-
-def test_property_type_hint():
-    prop = m.class_with_enum.__dict__["value"]
-    assert isinstance(prop, property)
-    assert prop.fget.__doc__.startswith(
-        "(self: pybind11_tests.native_enum.class_with_enum) -> pybind11_tests.native_enum.class_with_enum.in_class"
-    )
-
-
-def test_function_signature():
-    assert m.function_with_native_enum.__doc__.startswith(
-        "function_with_native_enum(n: pybind11_tests.native_enum.Native) -> pybind11_tests.native_enum.Native"
-    )

--- a/tests/test_native_enum.py
+++ b/tests/test_native_enum.py
@@ -139,7 +139,7 @@ def test_pass_color_success():
 def test_pass_color_fail():
     with pytest.raises(TypeError) as excinfo:
         m.pass_color(None)
-    assert "test_native_enum::color" in str(excinfo.value)
+    assert "pybind11_tests.native_enum.color" in str(excinfo.value)
 
 
 def test_return_color_success():
@@ -301,3 +301,9 @@ def test_native_enum_missing_finalize_failure():
     if not isinstance(m.native_enum_missing_finalize_failure, str):
         m.native_enum_missing_finalize_failure()
         pytest.fail("Process termination expected.")
+
+
+def test_function_signature():
+    assert m.function_with_native_enum.__doc__.startswith(
+        "function_with_native_enum(n: pybind11_tests.native_enum.Native) -> pybind11_tests.native_enum.Native"
+    )

